pkg/koordlet/statesinformer/states_pod.go

```diff
@@ -17,20 +17,29 @@ limitations under the License.
 package statesinformer
 
 import (
+	"context"
 	"time"
 
 	"golang.org/x/time/rate"
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/klog/v2"
 
 	"github.com/koordinator-sh/koordinator/pkg/util"
 )
 
-func (s *statesInformer) syncKubelet() error {
+func (s *statesInformer) syncPods() error {
 	podList, err := s.kubelet.GetAllPods()
-	if err != nil {
-		klog.Warningf("get pods from kubelet failed, err: %v", err)
-		return err
+
+	// when kubelet recovers form crash, podList may be empty.
+	if err != nil || len(podList.Items) == 0 {
+		klog.Warningf("get pods from kubelet failed back to apiserver, err: %v", err)
+		podList, err = s.getPodsFromApiserver()
```

Comment:
Should the getting apiserver be the default behavior for the fallback, rather than an option of `syncPods`? WDYT @LambdaHJ @jasonliu747 @zwzhang0107 @hormes

Reply:
Koordlet is a DeamonSet, so it seems to me instead of letting each daemon query APIServer, accessing Kubelet on each node might be a best choice to achieve better performance and scalability.
I'd love to hear your view on this PR as well @eahydra.

Reply:
According to the implementation of apiserver in the community version, requests received when APIServer itself failover will penetrate to etcd. Considering that koordlet itself runs in the form of a daemonset, it may generate a large number of requests to the APIServer under a large cluster scale, which affects the reliability of the cluster dimension. Therefore, we should consider protecting the APIServer. In addition, I am wondering if there are users who expect to go directly to the apiserver to synchronize data?

Reply:
Because the frequency of this call is very high, it falls back to calling the apiserver when it fails, which may lead to excessive pressure on the apiserver. A solution that can be considered is to use a list-watch as one of data source Of course, this can be implemented in another PR.